File: day06/scala/part1.scala
import io.StdIn.readLine
import util.Try
import util.chaining._

def getInput: String = Iterator
    .continually(readLine)
    .takeWhile(_ != null)
    .mkString("\n")

type Operator = Long => Long => Long

def sToOp(s: String): Operator = s match
    case "+" => (a) => (b) => a + b
    case "*" => (a) => (b) => a * b
    case _ => throw RuntimeException("unknown operator" + s)

def parse(s: String): Array[(Array[Long], Operator)] =
    s.linesIterator.map(line => line.split(' ').filter(_ != "")).toArray.transpose.map(p =>
        val len = p.length
        val numbers = p.take(len - 1)
        (numbers.map(s => util.Try(s.toLong).get), sToOp(p(len - 1)))
    )

def solve(arr: Array[(Array[Long], Operator)]): Long =
    arr.map((nums, op) => nums.reduce(Function.uncurried(op))).sum

@main def main() = println(getInput.pipe(parse).pipe(solve))

File: day06/hs/part1.hs
module Main where

import Data.List (transpose)

type Operator = Int -> Int -> Int

sToOp :: String -> Operator
sToOp "+" = (+)
sToOp "*" = (*)
sToOp _ = error "parse error"

parse :: String -> [([Int], Operator)]
parse = map parseLine . transpose . map words . lines
  where
    parseLine :: [String] -> ([Int], Operator)
    parseLine l = ((map read . take (len - 1)) l, (sToOp . last) l)
      where
        len = length l

solve :: [([Int], Operator)] -> Int
solve = sum . map foldingFn
  where
    foldingFn :: ([Int], Operator) -> Int
    foldingFn (l, op) = foldr1 op l

main :: IO ()
main = interact ((++ "\n") . show . solve . parse)
